### PR TITLE
Embedded-boundary thermal (diffuse) particle re-emission

### PR DESCRIPTION
**Pull request by clarkse-he**
TITLE: Embedded-boundary thermal (diffuse) particle re-emission

BASE: development   HEAD: Helion-Energy:dsmc-eb-reflection

## Summary
Adds **thermal (fully diffuse) re-emission** of particles at embedded boundaries, applied
uniformly to all species: `boundary.particle_eb = thermal`. Stock WarpX absorbs
(default) or specularly reflects (#6588) particles at an EB; for neutral-gas / DSMC
problems (gas expanding against solid walls) the wall must conserve the gas inventory
while thermally accommodating it.

Per review discussion below, this PR is scoped to the uniform-across-species diffuse
wall only, consistent with the `boundary.particle_eb = absorbing | reflecting`
interface introduced by #6588. Deferred to follow-up PRs, where they can be designed
coherently:
- the **per-species boundary selection and the accommodation coefficient** (mixed
  specular/diffuse walls) — preserved on `Helion-Energy:dsmc-eb-per-species-followup`;
- the **corner-case safeguards** (absorb on ill-defined wall interactions), since they
  affect the existing `reflecting` mode as well — preserved on
  `Helion-Energy:dsmc-eb-corner-safeguards`.

## Interface
- `boundary.particle_eb = thermal` — every particle striking the EB is re-emitted from
  a wall Maxwellian: half-Maxwellian flux along the inward wall normal, full Maxwellian
  tangentially (the same sampling primitives as the domain `thermal` particle boundary).
- `boundary.<species_name>.u_th = <sqrt(kB*T_wall/m)/c>` — wall thermal speed per
  species, the same input the domain thermal boundary condition uses.

## What's added
- The thermal re-emission is folded into #6588's reflection functor, renamed
  `ParticleBoundaryProcess::ParticleBoundaryInteraction` (as suggested in review,
  following #7058): bisect the trajectory to the exact wall-contact point, evaluate the
  level-set normal there (post-#7051 `interp_normal`, unit 3D Cartesian normal in every
  dimensionality), then either reflect specularly or re-emit from the wall Maxwellian —
  an isotropic 3D Gaussian draw whose normal component is replaced by a `gaussianflux`
  draw along the inward normal — and advance the particle for the remaining fraction of
  the step.
- Dispatch in `MultiParticleContainer::ScrapeParticlesAtEB` (Thermal alongside
  Reflecting, same `scrapeParticlesAtEB` call); re-emitted particles go through the full
  `Redistribute()` path in `WarpX::HandleParticlesAtBoundaries` (from #6588).
- `boundary.<species>.u_th` parsing is shared with the domain thermal BC
  (`isAnyParticleBoundaryThermal()` now also covers the EB;
  `getBoundaryThermalVelocity()` reuses `m_boundary_conditions.data.m_uth`).

## Test (`Examples/Tests/particle_boundary_interaction/`)
- `test_2d_particle_boundary_interaction_thermal` — `boundary.particle_eb = thermal`,
  600 K wall / 300 K gas in a field-free, collisionless closed box over a solid EB slab
  (isolating the wall model, ~1 s); asserts number conservation and thermalization
  toward (but not past) the wall temperature, plus the standard checksum regression on
  the final-step particle data.

Specular reflection at the EB is already covered by the existing
`test_rz_particle_boundary_interaction_reflecting`, so this PR no longer adds a
separate specular test.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01FyFeNNpr5jir3ygSakbZCm

### CODE REVIEW

File: Source/EmbeddedBoundary/ParticleReflectAtEB.H [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+                    xp, yp, zp, plo, dxi, i, j, k, W);
+                amrex::Real const phi_value = ablastr::particles::interp_field_nodal(i, j, k, W, phi);
+                if (phi_value >= 0.0_rt) { return; }  // particle still in the gas region
+
+                // Corner safeguard (a): if the particle penetrated deeper than the signed-distance
+                // field is reliable (~1 cell), it entered at a sharp corner or tunneled -> absorb it
+                // (otherwise it can accumulate in the solid over long runs / complex geometry).
+                amrex::Real dx_max = amrex::Real(1.0)/dxi[0];
+#if (AMREX_SPACEDIM >= 2)
+                dx_max = amrex::max(dx_max, amrex::Real(1.0)/dxi[1]);
```
Comment (clarkse-he) [inline Source/EmbeddedBoundary/ParticleReflectAtEB.H]: I think there is a macro expansion for selecting the Z Index to avoid this CPP split.

**Review by clarkse-he — COMMENTED** [review]
Looks pretty good.

**Comment by clarkse-he** [comment]
Following the review discussion in #7021 (https://github.com/BLAST-WarpX/warpx/pull/7021#issuecomment-4905689951), the reflection now uses the same bisection approach as #6588 and `ParticleBoundaryBuffer::FindEmbeddedBoundaryIntersection`, instead of mirroring the position across the local tangent plane:

- bisect along the particle trajectory to find the exact contact point where `phi = 0`;
- compute the surface normal at that contact point (correct for curved EB surfaces, and no longer leaves the particle inside the solid for the remainder of the step);
- apply the specular reflection or diffuse (Maxwell accommodation) re-emission at the contact point;
- advance the particle from there with its new velocity for the remaining fraction of the time step.

The previous deep-penetration heuristic is replaced by an exact check that the particle started the step in the gas region (if it did not, there is no crossing to bracket, and the particle is absorbed as before).

This also addresses the second point from that comment: when any species reflects at the EB, `WarpX::HandleParticlesAtBoundaries` now calls `Redistribute()` instead of `deleteInvalidParticles()`, since a reflected particle can end up owned by a different sub-domain.

The 2D specular and diffuse EB reflection tests pass with the new scheme.

**Comment by RemiLehe** [comment]
Thanks for this PR ; this looks great!

Before diving into the review, I had a few high-level questions about how this interface might be extended in the future. While these questions go beyond the current PR, I am asking them now in order to make sure that the current user interface would be easily extensible, so we don't need to do breaking changes to the user interface in the future.

- Right now, the option are either to absorb, or to reflection/diffuse with an accomodation coefficient to choose between reflection and diffusion. In the future, do think there might be a need to have (for a given species) a non-zero probability for **all 3 processes** (i.e. absorb/reflect/diffuse), and if yes what might be a good user interface for this? 

- In the future, could we have a need for a an energy-dependent (or velocity-dependent) accommodation coefficient?

**Comment by clarkse-he** [comment]
> Thanks for this PR ; this looks great!
> 
> Before diving into the review, I had a few high-level questions about how this interface might be extended in the future. While these questions go beyond the current PR, I am asking them now in order to make sure that the current user interface would be easily extensible, so we don't need to do breaking changes to the user interface in the future.
> 
> * Right now, the option are either to absorb, or to reflection/diffuse with an accomodation coefficient to choose between reflection and diffusion. In the future, do think there might be a need to have (for a given species) a non-zero probability for **all 3 processes** (i.e. absorb/reflect/diffuse), and if yes what might be a good user interface for this?
> * In the future, could we have a need for a an energy-dependent (or velocity-dependent) accommodation coefficient?

I certainly think for plasma models we will need an energy selection for absorption. In the low energy limit, energy dep reflection from quantum tunneling effects can be included. In a classical sense the energy selection would be useful for having something like a subgrid virtual cathode for SCL emission.

To be honest I have been thinking a bit about this, but wanted to hold off on updating the functor for that in the future. I think in the most general sense we could use an implicit analytical function or tabulated values to compute the reflection/absorption coefficient normal energy component of the particle wrt the wall. Perhaps if we decide on an interface I can add it to the functor then assert guard it for future work. Does this sound like a plan to you?

**Comment by RemiLehe** [comment]
Sounds good.

Regarding the user interface, I think that the current PR is actually doing two things at the same time:
- Adding options for diffuse reflection
- Having species-specific boundary condition (e.g. `<species_name>.eb_reflect`) 

I would suggest that this PR (7008) focuses on the first point (Adding options for diffuse reflection), and have it applied uniformly across all species (similarly to how #6588 already applies the reflection on the EB uniformly to all species with `boundary.particle_eb = reflecting`), while postponing the implementation of species-specific boundary condition to a subsequent PR. This is so that, in this subsequent PR, we can think more about a generic, consistent user interface for species-specific boundary conditions that would be applicable not only to the EB but also to boundaries of the domain. (e.g. enabling one species to be reflected at the lower boundary of the simulation domain, while another species would be absorbed at the same boundary - which right now is not possible in WarpX)

If you agree with this suggestion and we focus on an interface that (for now) applies uniformly to all species, here is a suggested interface that would be consistent with #6588 as well as with the existing user interface for the domain boundary condition (https://warpx.readthedocs.io/en/latest/usage/parameters.html#boundary.particle_lo-hi):

The options would be:
- `boundary.particle_eb = absorbing` (already introduced in #7008)
- `boundary.particle_eb = reflecting` (already introduced in #7008)
- `boundary.particle_eb = thermal`, `boundary.<species_name>.u_th = 0.1`(new, but consistent with https://warpx.readthedocs.io/en/latest/usage/parameters.html#boundary.particle_lo-hi ; corresponds to `accommodation=1`)

- `boundary.particle_eb = mixed`, `boundary.particle_eb.p_thermal_reemission = 0.25` (probability of reemission ; corresponds to the accomodation coefficient), `boundary.<species_name>.u_th = 0.1`

In the future, we could also enable `boundary.particle_eb.p_thermal_reemission(x, y, z, E)` to give a parser expression that depends on the spatial coordinates and the energy of the incident particles. This would match the syntax for many other WarpX parameters which can either be given as a constant, or a parser expression.

**Comment by clarkse-he** [comment]
Done — the PR is now scoped as you suggested.

- **Per-species selection and the accommodation coefficient are stripped out** of this PR
  (preserved on `Helion-Energy:dsmc-eb-per-species-followup` for the follow-up, where we can
  design the generic per-species boundary interface for domain faces + EB together).
- This PR now adds exactly one option, uniform across species and consistent with #6588:
  `boundary.particle_eb = thermal` with `boundary.<species_name>.u_th` — your suggested
  syntax, reusing the domain thermal BC's input (same name, units, and half-Maxwellian
  sampling primitives).
- Since #6588 merged, I also rebased onto development and dropped the parts it superseded:
  specular reflection is now #6588's `ParticleBoundaryProcess::Reflect`, and the
  `Redistribute()` handling in `HandleParticlesAtBoundaries` is yours. The thermal functor
  was ported to the post-#7051 `interp_normal` (the unit 3D Cartesian normal collapsed the
  old per-dimensionality emission branches into one block). The corner safeguards
  (absorb on started-in-solid / degenerate normal / re-entered solid) are kept.
- Tests updated to the uniform interface: the specular test now exercises
  `boundary.particle_eb = reflecting` and the thermal test `= thermal`; both pass
  (N and E conserved to machine precision / gas thermalizes 300 K → 548 K toward the
  600 K wall).

The `mixed` mode (`p_thermal_reemission`, i.e. the accommodation coefficient) and the
energy-dependent variant will come with the follow-up PR as discussed.

**Comment by RemiLehe** [comment]
Thanks for updating the PR!

Regarding the actual C++ implementation, for the sake of conciseness of code, I would suggest to use the same code as here:
https://github.com/BLAST-WarpX/warpx/pull/7058/changes#diff-d14770972e224cd2fac33412bb3342e89c144faf8dff83b91fb8e4df986799c5R131
i.e. fold the thermal emission into the existing `Reflect` functor (and rename this functor), and generate the random velocity by first drawing from an isotropic 3D Gaussian, and then replacing the normal component using the `gaussianflux` distribution (this avoids somewhat long code to generate a 3D orthonormal basis).

**Comment by clarkse-he** [comment]
Done — the implementation now follows #7058:

- The thermal emission is folded into the (renamed) `ParticleBoundaryProcess::ParticleBoundaryInteraction` functor, dispatched through the same `scrapeParticlesAtEB` call for both `reflecting` and `thermal`. The separate `ParticleReflectAtEB.H` is deleted.
- The re-emitted velocity is generated your way: an isotropic 3D Gaussian draw, with the normal component replaced by a `gaussianflux` draw along the inward normal — the orthonormal-basis construction is gone.
- Also adopted from #7058: `getBoundaryThermalVelocity()` reuses `m_boundary_conditions.data.m_uth` instead of a duplicate member, and `isAnyParticleBoundaryThermal()` now covers the EB so the `u_th` parsing is back to a single condition.

I kept the three corner safeguards, now inside the functor (absorb when the particle did not start the step in the gas region, when the level-set normal is degenerate at a sharp corner, or when the particle lands back inside the solid at a concave corner). One consequence to be aware of: since the functor is shared, these safeguards now protect `reflecting` as well as `thermal` — previously a specular reflection at a degenerate corner could produce a NaN momentum, and a particle starting inside the solid gave the bisection nothing to bracket. Happy to make them thermal-only if you'd rather leave `reflecting` behavior untouched in this PR.

Both 2D EB tests pass with the new sampling (particle number conserved exactly; the gas thermalizes 300 K → 541 K toward the 600 K wall).

**Comment by RemiLehe** [comment]
Thanks!

Since the 3 corner cases are now common to both reflection and diffuse re-emission, and since they might require more discussion, could you remove them from this PR and add them in a separate, follow-up PR?

**Comment by clarkse-he** [comment]
Done — the three corner-case safeguards are removed from this PR, so the shared functor
now handles corner cases exactly as the pre-existing `Reflect` functor did, for both
`reflecting` and `thermal`.

The safeguards (absorb when the particle did not start the step in the gas region, when
the level-set normal is degenerate at a sharp corner, or when the particle lands back
inside the solid at a concave corner) are preserved as a single clean commit on
`Helion-Energy:dsmc-eb-corner-safeguards`, and I will open the follow-up PR for that
discussion once this one merges.

Both 2D EB tests (specular and thermal) pass with the safeguards removed — the test
geometry is a flat slab, so no corner case is exercised.

File: Examples/Tests/embedded_boundary_reflection/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+#
+
+if(WarpX_EB)
+    add_warpx_test(
+        test_2d_eb_diffuse_reflection  # name
+        2  # dims
+        1  # nprocs
+        inputs_test_2d_eb_diffuse_reflection  # inputs
+        "analysis_eb_diffuse_reflection.py"  # analysis
+        OFF  # checksum
```
Comment (RemiLehe) [inline Examples/Tests/embedded_boundary_reflection/CMakeLists.txt]: The convention in WarpX is to generate checksums for all CI tests (see other tests and https://warpx.readthedocs.io/en/latest/developers/how_to_test.html#how-to-add-automated-tests)

Could you add the checksum analysis script (it should be a link to the standard script) and the corresponding JSON files that store checksum values?

File: Examples/Tests/embedded_boundary_reflection/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+        OFF  # dependency
+    )
+
+    add_warpx_test(
+        test_2d_eb_specular_reflection  # name
```
Comment (RemiLehe) [inline Examples/Tests/embedded_boundary_reflection/CMakeLists.txt]: I would remove this test from the PR: the PR is only about the thermal emission. Specular reflection is already tested here: https://github.com/BLAST-WarpX/warpx/blob/development/Examples/Tests/particle_boundary_interaction/inputs_test_rz_particle_boundary_interaction_reflecting

For similar reasons, could you move these tests to the existing folder `Examples/Tests/particle_boundary_interaction`, instead of creating the separate folder `embedded_boundary_reflection`?

File: Examples/Tests/embedded_boundary_reflection/inputs_test_2d_eb_diffuse_reflection [hunk 4]
```diff
@@ -0,0 +1,61 @@
+# 2D embedded-boundary neutral-particle reflection test -- THERMAL (boundary.particle_eb = thermal).
+#
+# Same field-free, collisionless geometry as the specular test (solid EB slab below z = 5 mm,
```
Comment (RemiLehe) [inline Examples/Tests/embedded_boundary_reflection/inputs_test_2d_eb_diffuse_reflection]: If we remove the specular test, this comment would need to change, so as not to refer to that test.

File: Examples/Tests/embedded_boundary_reflection/CMakeLists.txt (hunk 2 above)
Comment (clarkse-he) [reply Examples/Tests/embedded_boundary_reflection/CMakeLists.txt]: Done in c7f9b735f — added the standard checksum command (link to `analysis_default_regression.py`) on a particles-only plotfile diag at the final step, and the corresponding benchmark JSON, generated locally with `CHECKSUM_RESET=ON ctest`. If the CI runners disagree with my machine's values I'll reset from the CI logs.

File: Examples/Tests/embedded_boundary_reflection/CMakeLists.txt (hunk 3 above)
Comment (clarkse-he) [reply Examples/Tests/embedded_boundary_reflection/CMakeLists.txt]: Done in c7f9b735f — the specular test is removed, and the thermal test now lives in `Examples/Tests/particle_boundary_interaction/`. I also renamed it to `test_2d_particle_boundary_interaction_thermal` to match the folder's naming convention and the `thermal` interface.

File: Examples/Tests/embedded_boundary_reflection/inputs_test_2d_eb_diffuse_reflection (hunk 4 above)
Comment (clarkse-he) [reply Examples/Tests/embedded_boundary_reflection/inputs_test_2d_eb_diffuse_reflection]: Done in c7f9b735f — the header comment now describes the setup on its own, without referring to the removed specular test.

**Review by RemiLehe — APPROVED** [review]
